2008-11-25  Corentin Le Gall <clegall@proformatique.com>

	pf_manager_misc.patch
	$Revision$
	$Date$
        Miscellaneous addenda in Manager Events
        - backport of "MeetMeList" Manager Command from 1.6 branch
        - backport of "Transfer" and "Masquerade" Manager Events from 1.6 branch
        - added misc paramters to MeetMeJoin event
        - added Data contents to the Dial event
        - added Application Name and Data to the Status event
        - added Uniqueid information to the ParkedCall events

Index: asterisk-1.4.22/apps/app_dial.c
===================================================================
--- asterisk-1.4.22.orig/apps/app_dial.c	2008-08-06 01:13:20.000000000 +0200
+++ asterisk-1.4.22/apps/app_dial.c	2008-11-28 09:57:51.000000000 +0100
@@ -387,16 +387,21 @@
 static void senddialevent(struct ast_channel *src, struct ast_channel *dst)
 {
 	/* XXX do we need also CallerIDnum ? */
-	manager_event(EVENT_FLAG_CALL, "Dial", 
-			   "Source: %s\r\n"
-			   "Destination: %s\r\n"
-			   "CallerID: %s\r\n"
-			   "CallerIDName: %s\r\n"
-			   "SrcUniqueID: %s\r\n"
-			   "DestUniqueID: %s\r\n",
-			   src->name, dst->name, S_OR(src->cid.cid_num, "<unknown>"),
-			   S_OR(src->cid.cid_name, "<unknown>"), src->uniqueid,
-			   dst->uniqueid);
+	manager_event(EVENT_FLAG_CALL, "Dial",
+                      "Source: %s\r\n"
+                      "Destination: %s\r\n"
+                      "Data: %s\r\n"
+                      "CallerID: %s\r\n"
+                      "CallerIDName: %s\r\n"
+                      "SrcUniqueID: %s\r\n"
+                      "DestUniqueID: %s\r\n",
+                      src->name,
+                      dst->name,
+                      src->data ? src->data : "",
+                      S_OR(src->cid.cid_num, "<unknown>"),
+                      S_OR(src->cid.cid_name, "<unknown>"),
+                      src->uniqueid,
+                      dst->uniqueid);
 }
 
 static struct ast_channel *wait_for_answer(struct ast_channel *in, struct dial_localuser *outgoing, int *to, struct ast_flags *peerflags, int *sentringing, char *status, size_t statussize, int busystart, int nochanstart, int congestionstart, int priority_jump, int *result)
Index: asterisk-1.4.22/apps/app_meetme.c
===================================================================
--- asterisk-1.4.22.orig/apps/app_meetme.c	2008-08-13 22:42:33.000000000 +0200
+++ asterisk-1.4.22/apps/app_meetme.c	2008-11-28 09:57:51.000000000 +0100
@@ -1696,11 +1696,21 @@
 
 	if (!sent_event) {
 		manager_event(EVENT_FLAG_CALL, "MeetmeJoin", 
-			      "Channel: %s\r\n"
-			      "Uniqueid: %s\r\n"
-			      "Meetme: %s\r\n"
-			      "Usernum: %d\r\n",
-			      chan->name, chan->uniqueid, conf->confno, user->user_no);
+                              "Channel: %s\r\n"
+                              "Uniqueid: %s\r\n"
+                              "Meetme: %s\r\n"
+                              "PseudoChan: %s\r\n"
+                              "Admin: %s\r\n"
+                              "Usernum: %d\r\n"
+                              "CallerIDnum: %s\r\n"
+                              "CallerIDname: %s\r\n",
+			      chan->name, chan->uniqueid,
+                              conf->confno,
+                              conf->chan->name,
+                              user->userflags & CONFFLAG_ADMIN ? "Yes" : "No",
+                              user->user_no,
+                              S_OR(user->chan->cid.cid_num, "<unknown>"),
+                              S_OR(user->chan->cid.cid_name, "<unknown>"));
 		sent_event = 1;
 	}
 
@@ -3019,6 +3029,83 @@
 	return meetmemute(s, m, 0);
 }
 
+static char mandescr_meetmelist[] =
+"Description: Lists all users in a particular MeetMe conference.\n"
+"MeetmeList will follow as separate events, followed by a final event called\n"
+"MeetmeListComplete.\n"
+"Variables:\n"
+"    *ActionId: <id>\n"
+"    *Conference: <confno>\n";
+
+static int action_meetmelist(struct mansession *s, const struct message *m)
+{
+	const char *actionid = astman_get_header(m, "ActionID");
+	const char *conference = astman_get_header(m, "Conference");
+	char idText[80] = "";
+	struct ast_conference *cnf;
+	struct ast_conf_user *user;
+	int total = 0;
+
+	if (!ast_strlen_zero(actionid))
+		snprintf(idText, sizeof(idText), "ActionID: %s\r\n", actionid);
+
+	if (AST_LIST_EMPTY(&confs)) {
+		astman_send_error(s, m, "No active conferences.");
+		return 0;
+	}
+
+	astman_send_ack(s, m, "Meetme user list will follow");
+
+	/* Find the right conference */
+	AST_LIST_LOCK(&confs);
+	AST_LIST_TRAVERSE(&confs, cnf, list) {
+		/* If we ask for one particular, and this isn't it, skip it */
+		if (!ast_strlen_zero(conference) && strcmp(cnf->confno, conference))
+			continue;
+
+		/* Show all the users */
+		AST_LIST_TRAVERSE(&cnf->userlist, user, list) {
+			total++;
+			astman_append(s,
+                                      "Event: MeetmeList\r\n"
+                                      "%s"
+                                      "Conference: %s\r\n"
+                                      "PseudoChan: %s\r\n"
+                                      "UserNumber: %d\r\n"
+                                      "CallerIDNum: %s\r\n"
+                                      "CallerIDName: %s\r\n"
+                                      "Channel: %s\r\n"
+                                      "Admin: %s\r\n"
+                                      "Role: %s\r\n"
+                                      "MarkedUser: %s\r\n"
+                                      "Muted: %s\r\n"
+                                      "Talking: %s\r\n"
+                                      "\r\n",
+                                      idText,
+                                      cnf->confno,
+                                      cnf->chan->name,
+                                      user->user_no,
+                                      S_OR(user->chan->cid.cid_num, "<unknown>"),
+                                      S_OR(user->chan->cid.cid_name, "<no name>"),
+                                      user->chan->name,
+                                      user->userflags & CONFFLAG_ADMIN ? "Yes" : "No",
+                                      user->userflags & CONFFLAG_MONITOR ? "Listen only" : user->userflags & CONFFLAG_TALKER ? "Talk only" : "Talk and listen",
+                                      user->userflags & CONFFLAG_MARKEDUSER ? "Yes" : "No",
+                                      user->adminflags & ADMINFLAG_MUTED ? "By admin" : user->adminflags & ADMINFLAG_SELFMUTED ? "By self" : "No",
+                                      user->talking > 0 ? "Yes" : user->talking == 0 ? "No" : "Not monitored");
+		}
+	}
+	AST_LIST_UNLOCK(&confs);
+	/* Send final confirmation */
+	astman_append(s,
+                      "Event: MeetmeListComplete\r\n"
+                      "EventList: Complete\r\n"
+                      "ListItems: %d\r\n"
+                      "%s"
+                      "\r\n", total, idText);
+	return 0;
+}
+
 static void *recordthread(void *args)
 {
 	struct ast_conference *cnf = args;
@@ -4840,6 +4927,7 @@
 	ast_cli_unregister_multiple(cli_meetme, ARRAY_LEN(cli_meetme));
 	res = ast_manager_unregister("MeetmeMute");
 	res |= ast_manager_unregister("MeetmeUnmute");
+	res |= ast_manager_unregister("MeetmeList");
 	res |= ast_unregister_application(app3);
 	res |= ast_unregister_application(app2);
 	res |= ast_unregister_application(app);
@@ -4863,10 +4951,12 @@
 	res |= load_config(0);
 
 	ast_cli_register_multiple(cli_meetme, ARRAY_LEN(cli_meetme));
-	res |= ast_manager_register("MeetmeMute", EVENT_FLAG_CALL, 
+	res |= ast_manager_register("MeetmeMute", EVENT_FLAG_CALL,
 				    action_meetmemute, "Mute a Meetme user");
-	res |= ast_manager_register("MeetmeUnmute", EVENT_FLAG_CALL, 
+	res |= ast_manager_register("MeetmeUnmute", EVENT_FLAG_CALL,
 				    action_meetmeunmute, "Unmute a Meetme user");
+	res |= ast_manager_register2("MeetmeList", EVENT_FLAG_CALL,
+                                     action_meetmelist, "List participants in a conference", mandescr_meetmelist);
 	res |= ast_register_application(app3, admin_exec, synopsis3, descrip3);
 	res |= ast_register_application(app2, count_exec, synopsis2, descrip2);
 	res |= ast_register_application(app, conf_exec, synopsis, descrip);
Index: asterisk-1.4.22/main/manager.c
===================================================================
--- asterisk-1.4.22.orig/main/manager.c	2008-11-28 09:57:51.000000000 +0100
+++ asterisk-1.4.22/main/manager.c	2008-11-28 09:57:51.000000000 +0100
@@ -1555,49 +1555,53 @@
 				elapsed_seconds = now.tv_sec - c->cdr->start.tv_sec;
 			}
 			astman_append(s,
-			"Event: Status\r\n"
-			"Privilege: Call\r\n"
-			"Channel: %s\r\n"
-			"CallerID: %s\r\n"		/* This parameter is deprecated and will be removed post-1.4 */
-			"CallerIDNum: %s\r\n"
-			"CallerIDName: %s\r\n"
-			"Account: %s\r\n"
-			"State: %s\r\n"
-			"Context: %s\r\n"
-			"Extension: %s\r\n"
-			"Priority: %d\r\n"
-			"Seconds: %ld\r\n"
-			"%s"
-			"Uniqueid: %s\r\n"
-			"%s"
-			"\r\n",
-			c->name, 
-			S_OR(c->cid.cid_num, "<unknown>"), 
-			S_OR(c->cid.cid_num, "<unknown>"), 
-			S_OR(c->cid.cid_name, "<unknown>"), 
-			c->accountcode,
-			ast_state2str(c->_state), c->context,
-			c->exten, c->priority, (long)elapsed_seconds, bridge, c->uniqueid, idText);
+                                      "Event: Status\r\n"
+                                      "Privilege: Call\r\n"
+                                      "Channel: %s\r\n"
+                                      "Application: %s\r\n"
+                                      "AppData: %s\r\n"
+                                      "CallerIDNum: %s\r\n"
+                                      "CallerIDName: %s\r\n"
+                                      "Account: %s\r\n"
+                                      "State: %s\r\n"
+                                      "Context: %s\r\n"
+                                      "Extension: %s\r\n"
+                                      "Priority: %d\r\n"
+                                      "Seconds: %ld\r\n"
+                                      "%s"
+                                      "Uniqueid: %s\r\n"
+                                      "%s"
+                                      "\r\n",
+                                      c->name,
+                                      c->appl ? c->appl : "",
+                                      c->data ? c->data : "",
+                                      S_OR(c->cid.cid_num, "<unknown>"),
+                                      S_OR(c->cid.cid_name, "<unknown>"),
+                                      c->accountcode,
+                                      ast_state2str(c->_state), c->context,
+                                      c->exten, c->priority, (long)elapsed_seconds, bridge, c->uniqueid, idText);
 		} else {
 			astman_append(s,
-			"Event: Status\r\n"
-			"Privilege: Call\r\n"
-			"Channel: %s\r\n"
-			"CallerID: %s\r\n"		/* This parameter is deprecated and will be removed post-1.4 */
-			"CallerIDNum: %s\r\n"
-			"CallerIDName: %s\r\n"
-			"Account: %s\r\n"
-			"State: %s\r\n"
-			"%s"
-			"Uniqueid: %s\r\n"
-			"%s"
-			"\r\n",
-			c->name, 
-			S_OR(c->cid.cid_num, "<unknown>"), 
-			S_OR(c->cid.cid_num, "<unknown>"), 
-			S_OR(c->cid.cid_name, "<unknown>"), 
-			c->accountcode,
-			ast_state2str(c->_state), bridge, c->uniqueid, idText);
+                                      "Event: Status\r\n"
+                                      "Privilege: Call\r\n"
+                                      "Channel: %s\r\n"
+                                      "Application: %s\r\n"
+                                      "AppData: %s\r\n"
+                                      "CallerIDNum: %s\r\n"
+                                      "CallerIDName: %s\r\n"
+                                      "Account: %s\r\n"
+                                      "State: %s\r\n"
+                                      "%s"
+                                      "Uniqueid: %s\r\n"
+                                      "%s"
+                                      "\r\n",
+                                      c->name,
+                                      c->appl ? c->appl : "",
+                                      c->data ? c->data : "",
+                                      S_OR(c->cid.cid_num, "<unknown>"),
+                                      S_OR(c->cid.cid_name, "<unknown>"),
+                                      c->accountcode,
+                                      ast_state2str(c->_state), bridge, c->uniqueid, idText);
 		}
 		ast_channel_unlock(c);
 		if (!all)
Index: asterisk-1.4.22/res/res_features.c
===================================================================
--- asterisk-1.4.22.orig/res/res_features.c	2008-11-28 09:57:51.000000000 +0100
+++ asterisk-1.4.22/res/res_features.c	2008-11-28 09:57:51.000000000 +0100
@@ -426,17 +426,23 @@
 		ast_verbose(VERBOSE_PREFIX_2 "Parked %s on %d@%s. Will timeout back to extension [%s] %s, %d in %d seconds\n", pu->chan->name, pu->parkingnum, parking_con, pu->context, pu->exten, pu->priority, (pu->parkingtime/1000));
 
 	manager_event(EVENT_FLAG_CALL, "ParkedCall",
-		"Exten: %s\r\n"
-		"Channel: %s\r\n"
-		"From: %s\r\n"
-		"Timeout: %ld\r\n"
-		"CallerID: %s\r\n"
-		"CallerIDName: %s\r\n",
-		pu->parkingexten, pu->chan->name, peer ? peer->name : "",
-		(long)pu->start.tv_sec + (long)(pu->parkingtime/1000) - (long)time(NULL),
-		S_OR(pu->chan->cid.cid_num, "<unknown>"),
-		S_OR(pu->chan->cid.cid_name, "<unknown>")
-		);
+                      "Exten: %s\r\n"
+                      "Channel: %s\r\n"
+                      "Uniqueid: %s\r\n"
+                      "From: %s\r\n"
+                      "UniqueidFrom: %s\r\n"
+                      "Timeout: %ld\r\n"
+                      "CallerID: %s\r\n"
+                      "CallerIDName: %s\r\n",
+                      pu->parkingexten,
+                      pu->chan->name,
+                      pu->chan->uniqueid,
+                      peer ? peer->name : "",
+                      peer ? peer->uniqueid : "",
+                      (long)pu->start.tv_sec + (long)(pu->parkingtime/1000) - (long)time(NULL),
+                      S_OR(pu->chan->cid.cid_num, "<unknown>"),
+                      S_OR(pu->chan->cid.cid_name, "<unknown>")
+                      );
 
 	if (peer && adsipark && ast_adsi_available(peer)) {
 		adsi_announce_park(peer, pu->parkingexten);	/* Only supports parking numbers */
@@ -1773,15 +1779,17 @@
 static void post_manager_event(const char *s, char *parkingexten, struct ast_channel *chan)
 {
 	manager_event(EVENT_FLAG_CALL, s,
-		"Exten: %s\r\n"
-		"Channel: %s\r\n"
-		"CallerID: %s\r\n"
-		"CallerIDName: %s\r\n\r\n",
-		parkingexten, 
-		chan->name,
-		S_OR(chan->cid.cid_num, "<unknown>"),
-		S_OR(chan->cid.cid_name, "<unknown>")
-		);
+                      "Exten: %s\r\n"
+                      "Channel: %s\r\n"
+                      "Uniqueid: %s\r\n"
+                      "CallerID: %s\r\n"
+                      "CallerIDName: %s\r\n\r\n",
+                      parkingexten,
+                      chan->name,
+                      chan->uniqueid,
+                      S_OR(chan->cid.cid_num, "<unknown>"),
+                      S_OR(chan->cid.cid_name, "<unknown>")
+                      );
 }
 
 /*! \brief Take care of parked calls and unpark them if needed */
@@ -2058,15 +2066,21 @@
 			ast_log(LOG_WARNING, "Whoa, no parking context?\n");
 
 		manager_event(EVENT_FLAG_CALL, "UnParkedCall",
-			"Exten: %s\r\n"
-			"Channel: %s\r\n"
-			"From: %s\r\n"
-			"CallerID: %s\r\n"
-			"CallerIDName: %s\r\n",
-			pu->parkingexten, pu->chan->name, chan->name,
-			S_OR(pu->chan->cid.cid_num, "<unknown>"),
-			S_OR(pu->chan->cid.cid_name, "<unknown>")
-			);
+                              "Exten: %s\r\n"
+                              "Channel: %s\r\n"
+                              "Uniqueid: %s\r\n"
+                              "From: %s\r\n"
+                              "UniqueidFrom: %s\r\n"
+                              "CallerID: %s\r\n"
+                              "CallerIDName: %s\r\n",
+                              pu->parkingexten,
+                              pu->chan->name,
+                              pu->chan->uniqueid,
+                              chan->name,
+                              chan->uniqueid,
+                              S_OR(pu->chan->cid.cid_num, "<unknown>"),
+                              S_OR(pu->chan->cid.cid_name, "<unknown>")
+                              );
 
 		free(pu);
 	}
Index: asterisk-1.4.22/channels/chan_sip.c
===================================================================
--- asterisk-1.4.22.orig/channels/chan_sip.c	2008-11-28 09:57:51.000000000 +0100
+++ asterisk-1.4.22/channels/chan_sip.c	2008-11-28 09:57:51.000000000 +0100
@@ -14798,6 +14798,19 @@
 	ast_set_flag(&transferer->flags[0], SIP_DEFER_BYE_ON_TRANSFER);	/* Delay hangup */
 
 	/* Perform the transfer */
+	manager_event(EVENT_FLAG_CALL, "Transfer",
+                      "TransferMethod: SIP\r\n"
+                      "TransferType: Attended\r\n"
+                      "Channel: %s\r\n"
+                      "Uniqueid: %s\r\n"
+                      "SIP-Callid: %s\r\n"
+                      "TargetChannel: %s\r\n"
+                      "TargetUniqueid: %s\r\n",
+                      transferer->owner->name,
+                      transferer->owner->uniqueid,
+                      transferer->callid,
+                      target.chan1->name,
+                      target.chan1->uniqueid);
 	res = attempt_transfer(current, &target);
 	ast_mutex_unlock(&targetcall_pvt->lock);
 	if (res) {
@@ -15064,6 +15077,22 @@
 		ast_clear_flag(&p->flags[0], SIP_GOTREFER);	
 		p->refer->status = REFER_200OK;
 		append_history(p, "Xfer", "REFER to call parking.");
+		manager_event(EVENT_FLAG_CALL, "Transfer",
+                              "TransferMethod: SIP\r\n"
+                              "TransferType: Blind\r\n"
+                              "Channel: %s\r\n"
+                              "Uniqueid: %s\r\n"
+                              "SIP-Callid: %s\r\n"
+                              "TargetChannel: %s\r\n"
+                              "TargetUniqueid: %s\r\n"
+                              "TransferExten: %s\r\n"
+                              "Transfer2Parking: Yes\r\n",
+                              current.chan1->name,
+                              current.chan1->uniqueid,
+                              p->callid,
+                              current.chan2->name,
+                              current.chan2->uniqueid,
+                              p->refer->refer_to);
 		if (sipdebug && option_debug > 3)
 			ast_log(LOG_DEBUG, "SIP transfer to parking: trying to park %s. Parked by %s\n", current.chan2->name, current.chan1->name);
 		sip_park(current.chan2, current.chan1, req, seqno);
@@ -15135,6 +15164,22 @@
 	res = ast_async_goto(current.chan2, p->refer->refer_to_context, p->refer->refer_to, 1);
 
 	if (!res) {
+		manager_event(EVENT_FLAG_CALL, "Transfer",
+                              "TransferMethod: SIP\r\n"
+                              "TransferType: Blind\r\n"
+                              "Channel: %s\r\n"
+                              "Uniqueid: %s\r\n"
+                              "SIP-Callid: %s\r\n"
+                              "TargetChannel: %s\r\n"
+                              "TargetUniqueid: %s\r\n"
+                              "TransferExten: %s\r\n"
+                              "TransferContext: %s\r\n",
+                              current.chan1->name,
+                              current.chan1->uniqueid,
+                              p->callid,
+                              current.chan2->name,
+                              current.chan2->uniqueid,
+                              p->refer->refer_to, p->refer->refer_to_context);
 		/* Success  - we have a new channel */
 		if (option_debug > 2)
 			ast_log(LOG_DEBUG, "%s transfer succeeded. Telling transferer.\n", p->refer->attendedtransfer? "Attended" : "Blind");
Index: asterisk-1.4.22/main/channel.c
===================================================================
--- asterisk-1.4.22.orig/main/channel.c	2008-11-28 10:21:29.000000000 +0100
+++ asterisk-1.4.22/main/channel.c	2008-11-28 10:22:16.000000000 +0100
@@ -3423,7 +3423,15 @@
 	if (option_debug > 3)
 		ast_log(LOG_DEBUG, "Actually Masquerading %s(%d) into the structure of %s(%d)\n",
 			clone->name, clone->_state, original->name, original->_state);
-
+        
+	manager_event(EVENT_FLAG_CALL, "Masquerade",
+                      "Clone: %s\r\n"
+                      "CloneState: %s\r\n"
+                      "Original: %s\r\n"
+                      "OriginalState: %s\r\n",
+                      clone->name, ast_state2str(clone->_state),
+                      original->name, ast_state2str(original->_state));
+        
 	/* XXX This is a seriously wacked out operation.  We're essentially putting the guts of
 	   the clone channel into the original channel.  Start by killing off the original
 	   channel's backend.   I'm not sure we're going to keep this function, because
